Adds various quality of life changes for cooking to make it less click intensive.

Ketrai opened the pull request:
## About The Pull Request

- Increases tray item size by 1 item.

- Ranges and griddles can now be fed from trays.

Click when closed => fill soup pot.
Click when open => fill associated oven tray.
Right click when open => fill tray from oven tray
Click griddle => fill griddle surface.
Right click => fill tray from griddle surface

- Martian batter is now 5u of each ingredient into 10u of batter.

Hopefully will make it bug out less where it makes far fewer reagents than it is supposed to, fixing reagents, or well soups specifically... is out of scope for this PR.

- Adds the ability to print soup pots and large trays from the service lathe

Soup pot: 5 Iron sheets, 0.4 bluespace crystal (given their size of 200U)
Large serving tray: 2 iron sheets

## Why It's Good For The Game

Makes cooking a lot less tedious. Especially for people with low precision when it comes to filling oven trays. This also bring the behavior up to parity with how you can click microwaves with trays to fill them, ditto for the food processor. It also allows chef to use the whole capacity of an oven, as previously you couldn't easily click 6 cake batters or other giant sprites onto the tiny tray.

The tray is now sized to be able to easily feed a griddle 8 items.

## Changelog

:cl:
qol: chef equipment can now deposit and withdraw to/from trays!
qol: chef now has access to griddle and oven sized trays!
qol: service can now print soup pots
/:cl:

Ketrai committed:
Cooking QoL proc updates

Ketrai commented:
Oh woops, accidentally pushed a config update, let me fix that

Ketrai committed:
Reverts config change

MrMelbert commented on code/modules/food_and_drinks/machinery/oven.dm:
```diff
@@ -243,6 +248,34 @@
 	max_items = 6
 	biggest_w_class = WEIGHT_CLASS_BULKY
 
+/obj/item/plate/oven_tray/attackby(obj/item/item, mob/living/user, params)
```
You should use `item_interaction` rather than `attackby`

MrMelbert commented on code/modules/food_and_drinks/machinery/oven.dm:
```diff
@@ -243,6 +248,34 @@
 
+/obj/item/plate/oven_tray/attackby(obj/item/item, mob/living/user, params)
+
+	if(istype(item, /obj/item/storage))
```
I think checking for `/obj/item/storage` directly isn't the best practice, rather you should check if the atom has an atom storage (ie, `.atom_storage`) and work off of that

MrMelbert commented on code/modules/food_and_drinks/machinery/stove.dm:
```diff
@@ -115,6 +115,25 @@
 	. = ..()
 	LAZYREMOVE(added_ingredients, gone)
 
+/obj/item/reagent_containers/cup/soup_pot/proc/containeradd(obj/item/attacking_item, mob/user, params)
```
Document this proc?

MrMelbert replied:
I think it could use a better name though, perhaps `transfer_from_container_to_pot`. 

And the args - rename `attacking_item` to `transfer_from`, and delete `params` (you don't pass it anyways)

MrMelbert commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -63,6 +63,29 @@
 
 
 /obj/machinery/griddle/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/storage))
```
Ditto here (use `item_interaction` and check for atom storage)

MrMelbert reviewed (commented):
I already added the ability to dump from storage onto griddles a few months back, not really sure why we need two methods of doing it now but Shrug

ophaq commented:
You know what I've been hoping to have done for a while that you might be able to include in this PR (only if you want to)?

1. Have oven tray items such as cakes and pastry bases be able to be scooped without the oven tray itself into a serving tray.
2. Be able to scoop things up with the oven tray?
3. Have griddles and ovens be able to be fed via conveyor belt (very important because I want semi automated cooking for a gimmick).

If you could find a way to implement these, that would be really cool.

Ketrai committed:
Makes oven trays use item_interaction instead

ophaq commented:
>     3. Sounds like something to try in another PR. As I'd want to add some stuff like making the ingredients lower quality if they've been conveyor fed.

If it's just something like enabling the griddle and oven to recognize that the food was placed onto it like a tray would have, I don't see the logic in making the food lower quality since you still have to watch the food and press the switches. But I get it.

Ketrai committed:
Use item interaction instead.

Ketrai committed:
Adds the ability to collect from griddles and trays

Ketrai committed:
Prevents trays from beating up ovens

Ketrai committed:
Rename containeradd to transfer_from_container_to_pot

Ketrai commented:
> I already added the ability to dump from storage onto griddles a few months back, not really sure why we need two methods of doing it now but Shrug

I thought about this as well, figured it makes more sense to make both possible

> 1. Have oven tray items such as cakes and pastry bases be able to be scooped without the oven tray itself into a serving tray.

Implemented this.

Ketrai committed:
Forgot to compile

Ketrai committed:
Fixes documentation

Wallemations commented:
Rather than making a new tray, why not just make the oven tray able to work like a serving tray?

Ketrai commented:
> Rather than making a new tray, why not just make the oven tray able to work like a serving tray?

- I've not yet seen any cooks remove the trays. They tend to stay in the oven for ease of use.
- Oven trays are plates, which hard to click and work with, given that the food items on top of them have priority for clicking/interacting. I don't want to mess with that code and affect that functionality just to make them easier to use outside of the oven or as glorified food displays, and their current behavior serves them well in the oven. Trays don't have this quirk, being better to move stuff around with in the kitchen, complete with 3 pickup modes like other storage bags.

Metekillot commented:
I really like this PR. Pixel size is the gate for chef cooking speed

Ketrai commented:
> I just think having two identical trays with the only difference being size is confusing.
I actually added a sprite for this. 
![image](https://github.com/tgstation/tgstation/assets/28861083/70e57c8f-1a29-489e-9be3-61fc2705c28f)
If need be I can update the description to say they can hold 6 and 8 items respectively?

MrMelbert commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,42 @@
+		. = ..()
+		return
+	
+	if(item.atom_storage && is_right_clicking)
```
We can do 
```dm
if(is_right_clicking) 
    ...
else
    ...
```
Rather than
```dm
if(!is_right_clicking) 
    ...
if(is_right_clicking)
    ...
```

MrMelbert commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,42 @@
+				AddToGrill(tray_item, user)
+		if(loaded)
+			to_chat(user, span_notice("You insert [loaded] items into \the [src]."))
+			update_appearance()
+		. = ..()
+		return
```
```suggestion
			update_appearance()
		return
```
- Because we call parent up at the top, we don't need this parent call here anymore. 
- And because we did something successfully, `return ITEM_INTERACT_SUCCESS`.

MrMelbert commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,42 @@
+				continue
+			if(contents.len >= max_items)
+				balloon_alert(user, "it's full!")
+				return TRUE
```
```suggestion
				return ITEM_INTERACT_BLOCKING
```

MrMelbert commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,42 @@
+			// Non-tray dumping requires a do_after
+			to_chat(user, span_notice("You start dumping out the contents of [item] into [src]..."))
+			if(!do_after(user, 2 SECONDS, target = tray))
+				return
```
```suggestion
				return ITEM_INTERACT_BLOCKING
```

MrMelbert commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,42 @@
+		for(var/obj/tray_item in griddled_objects)
+			tray.atom_storage?.attempt_insert(tray_item, user, TRUE)
+	
+	. = ..()
+	return
```
Likewise we can delete this stuff and `return ITEM_INTERACT_SUCCESS` instead

EricZilla commented:
Wouldn't it make sense to make the oven tray just also be a large serving tray? No one removes it but that's not a reason to have a duplicate item the same size and same functions.

Ketrai committed:
Update code/game/objects/items/storage/bags.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Ketrai committed:
Update code/game/objects/items/storage/bags.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Ketrai commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,42 @@
 	else
 		return ..()
 
+/obj/machinery/griddle/item_interaction(mob/living/user, obj/item/item, list/modifiers, is_right_clicking)
+
+	if(item.atom_storage && !is_right_clicking)
```
Good to know, thanks for that info!!

Ketrai commented on code/modules/food_and_drinks/machinery/stove_component.dm:
```diff
@@ -128,6 +128,14 @@
 /datum/component/stove/proc/on_attackby(obj/machinery/source, obj/item/attacking_item, mob/user, params)
 	SIGNAL_HANDLER
 
+	if(istype(source, /obj/machinery/oven/range) && istype(attacking_item, /obj/item/storage/bag/tray) && container)
+		var/obj/machinery/oven/range/range = source
+		var/obj/item/reagent_containers/cup/soup_pot/soup_pot = container
+
+		if(!range.open)
+			soup_pot.transfer_from_container_to_pot(attacking_item, user)
+			return COMPONENT_NO_AFTERATTACK
```
I figured it's easier if you can just click the range itself, less precision needed, faster cooking. But you're absolutely right in that I should add the option to just click the soup pot directly too for people that wish to add items to it on a counter or something.

MrMelbert replied:
IMO you should just have it be click on the soup pot and not on the range, that way clicking on the stove itself has a defined behavior (attempt to load the tray) and clicking on the pot has its own (attempt to load the pot)

Ketrai committed:
Apply suggestions from code review

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Ketrai commented:
> Wouldn't it make sense to make the oven tray just also be a large serving tray? No one removes it but that's not a reason to have a duplicate item the same size and same functions.
> 
> Just stick an extra one in the starting kitchen so they figure it out. Rename oven tray to large tray, make small tray go in da oven, now the only difference is the size

As stated before they have different functions.
Oven tray : Can be clicked to remove items from. This makes them convenient and intuitive for oven usage, but annoying as actual trays, as you might click the item on top of the tray instead of the tray itself to move it.
Tray: Works as a storage item, click it to open. Will dump contents on any table clicked, unlike trays.

It works pretty great as is, in my opinion.

Ketrai committed:
Applies proper item interaction usage

EricZilla commented:
Ah okay I misunderstood what was going on

Wallemations commented:
> > I just think having two identical trays with the only difference being size is confusing.
> 
> 
> 
> I actually added a sprite for this. 
> 
> 
> 
> ![image](https://github.com/tgstation/tgstation/assets/28861083/70e57c8f-1a29-489e-9be3-61fc2705c28f)
> 
> If need be I can update the description to say they can hold 6 and 8 items respectively?
> 
> 

I didn't mean identical in visuals, I meant in function. Their only real functional difference is their capacity, correct? Sure writing in the desc their amounts could work but realistically people would just spring for the larger one without thinking because it has more overall utility due to its capacity.

Ketrai commented:
> I didn't mean identical in visuals, I meant in function. Their only real functional difference is their capacity, correct? Sure writing in the desc their amounts could work but realistically people would just spring for the larger one without thinking because it has more overall utility due to its capacity.
> 
> I think it would be a good idea to ask a code maint how hard it would be to change the oven tray's function itself rather than adding a new tray. If you're worried about its stacking then you could just make it force tray-loaded items into a grid

If it's that big of an issue I can cave for making them hold 8.
- Just make the default tray hold 8 and have chefs live with the extra click needed to put excess ingredients back onto the table.

I'm pretty certain people experienced with the cooking system would use the smaller one to feed ovens after like making a pile of raw pastry base for example. Scoop the pile, click the oven, drop the tray.

Changing the oven tray's functionality is difficult. From what I can garner from the code they sorta just attach items to themselves when they get added, but the items remain fully interactive and get removed from the plate when the item itself is moved, not when the plate is interacted with. The easiest solution here would be to just make it a tray subtype and have it be able to be right clicked with item interaction to remove the last item that was added.

Ketrai committed:
removes the large tray

Wallemations commented:
That's a way simpler fix than my proposed one, nice work!

jlsnow301 commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,45 @@
+/obj/machinery/griddle/item_interaction(mob/living/user, obj/item/item, list/modifiers, is_right_clicking)
+	. = ..()
+	if(. & ITEM_INTERACT_ANY_BLOCKER)
+		return .
```
`.` is implied,

```suggestion
		return
```

jlsnow301 replied:
(same applies for other instances of `return .`)

jlsnow301 commented on code/modules/food_and_drinks/machinery/oven.dm:
```diff
@@ -101,13 +101,22 @@
 
 
 /obj/machinery/oven/attackby(obj/item/I, mob/user, params)
-	if(open && !used_tray && istype(I, /obj/item/plate/oven_tray))
+	//if(open && used_tray && I.atom_storage)
+		//Stop containers from beating up our oven in general.
+		//return
+
+	if(open && !used_tray && I.atom_storage)
```
what happened here? why is this code commented out? can this just be early returned?

also, please, give `I` a better name, like `weapon` or something

Ketrai replied:
Woops, forgot to delete that comment. I was the original name, but I'll change it, magic numbers bad.

jlsnow301 commented on code/modules/food_and_drinks/machinery/oven.dm:
```diff
@@ -243,6 +252,43 @@
+	if(isnull(item.atom_storage))
+		return .
+
+	if(!is_right_clicking)
```
pretty big if block here, would be a better choice to early return `is_right_clicking`

ex
```
	if(is_right_clicking)
		var/obj/item/storage/tray = item

		for(var/obj/tray_item in contents)
			tray.atom_storage?.attempt_insert(tray_item, user, TRUE)

		return ITEM_INTERACT_SUCCESS
// rest of code can be de-indented
```

jlsnow301 commented on code/modules/food_and_drinks/machinery/stove.dm:
```diff
@@ -115,6 +115,32 @@
+ * * user: the entity adding ingredients via a container to a soup pot. Must not be null.
+ */
+/obj/item/reagent_containers/cup/soup_pot/proc/transfer_from_container_to_pot(obj/item/transfer_from, mob/user)
+	if(transfer_from.atom_storage)
```
early return this

jlsnow301 commented on code/modules/food_and_drinks/machinery/oven.dm:
```diff
@@ -243,6 +252,43 @@
+	. = ..()
+
+	if(isnull(item.atom_storage))
+		return .
```
```suggestion
		return
```

Ketrai committed:
Apply suggestions from code review

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

Ketrai committed:
Applies maintainer requests

jlsnow301 commented on code/modules/food_and_drinks/machinery/oven.dm:
```diff
@@ -100,15 +100,11 @@
-			to_chat(user, span_notice("You put [I] in [src]."))
-			add_tray_to_oven(I, user)
+/obj/machinery/oven/attackby(obj/item/item, mob/user, params)
+	if(open && !used_tray && item.atom_storage)
+		if(user.transferItemToLoc(item, src, silent = FALSE))
+			to_chat(user, span_notice("You put [item] in [src]."))
+			add_tray_to_oven(item, user)
 	else
 		return ..()
```
```suggestion
	if(!open || used_tray || !item.atom_storage)
		return ..()
		
	if(user.transferItemToLoc(item, src, silent = FALSE))
		to_chat(user, span_notice("You put [item] in [src]."))
		add_tray_to_oven(item, user)
```

jlsnow301 commented on code/modules/food_and_drinks/machinery/griddle.dm:
```diff
@@ -79,6 +80,44 @@
+		return
+	
+	if(isnull(item.atom_storage))
+		return .
```
```suggestion
		return
```

Ketrai committed:
Apply suggestions from code review

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>
